typescript/agentkit/src/wallet-providers/privyEvmEmbeddedWalletProvider.ts

```diff
@@ -0,0 +1,517 @@
+      method: "eth_signTransaction",
+      params: {
+        transaction: {
+          ...transaction,
+          from: this.#address,
+        },
+      },
+    };
+
+    try {
+      const response =
+        await this.executePrivyRequest<PrivyResponse<{ signed_transaction: Hex }>>(body);
+      return response.data?.signed_transaction;
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(`Transaction signing failed: ${error.message}`);
+      }
+      throw new Error("Transaction signing failed");
+    }
+  }
+
+  /**
+   * Sends a transaction.
+   *
+   * @param transaction - The transaction to send.
+   * @returns The hash of the transaction.
+   */
+  async sendTransaction(transaction: TransactionRequest): Promise<Hex> {
+    const body = {
+      address: this.#address,
+      chain_type: "ethereum",
+      method: "eth_sendTransaction",
+      params: {
+        transaction: {
+          ...transaction,
+          from: this.#address,
+        },
+      },
+    };
+
+    try {
+      const response = await this.executePrivyRequest<PrivyResponse<{ hash: Hex }>>(body);
+      return response.data?.hash;
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(`Transaction sending failed: ${error.message}`);
+      }
+      throw new Error("Transaction sending failed");
+    }
+  }
+
+  /**
+   * Waits for a transaction receipt.
+   *
+   * @param txHash - The hash of the transaction to wait for.
+   * @returns The transaction receipt.
+   */
+  async waitForTransactionReceipt(txHash: Hex): Promise<TransactionReceipt> {
+    return await this.#publicClient.waitForTransactionReceipt({
+      hash: txHash,
+    });
+  }
+
+  /**
+   * Reads data from a smart contract.
+   *
+   * @param params - Parameters for reading the contract
+   * @param params.address - The address of the contract
+   * @param params.abi - The ABI of the contract
+   * @param params.functionName - The name of the function to call
+   * @param params.args - The arguments to pass to the function
+   * @returns A Address that resolves to the contract function's return value
+   */
+  async readContract<
+    const abi extends Abi | readonly unknown[],
+    functionName extends ContractFunctionName<abi, "pure" | "view">,
+    const args extends ContractFunctionArgs<abi, "pure" | "view", functionName>,
+  >(
+    params: ReadContractParameters<abi, functionName, args>,
+  ): Promise<ReadContractReturnType<abi, functionName, args>> {
+    const body = {
+      address: this.#address,
+      chain_type: "ethereum",
+      chain_id: NETWORK_ID_TO_CHAIN_ID[this.#network.chainId!],
+      contract: {
+        address: params.address,
+        abi: params.abi,
+        functionName: params.functionName,
+        args: params.args,
+      },
+    };
+
+    const response = await this.executePrivyRequest<{
+      result: ReadContractReturnType<abi, functionName, args>;
+    }>({
+      method: "eth_call",
```

Comment:
I'm unable to read contract details. Debugging the error from privy, I get this internal response: 
> [Input error] `method`: Invalid discriminator value. Expected 'eth_signTransaction' | 'eth_sendTransaction' | 'personal_sign' | 'eth_signTypedData_v4' or `method`: Invalid discriminator value. Expected 'signTransaction' | 'signAndSendTransaction' | 'signMessage'

For whatever reason, it looks like `eth_call` is not a supported method.

You do have a publicClient instantiated from Viem, so I would recommend doing the read through it. It would also reduce code complexity

```ts
return this.#publicClient.readContract<abi, functionName, args>(params);
```

Reply:
ah I see. that makes sense! works now